src/render/graph.rs

```diff
@@ -1,190 +1,22 @@
-        // if the cpal buffer size did not align with our internal RENDER_QUANTUM_SIZE
-        if let Some((offset, prev_rendered)) = self.buffer_offset.take() {
-            let leftover_len = (RENDER_QUANTUM_SIZE - offset) * self.channels;
-            // split the leftover frames slice, to fit in `buffer`
-            let (first, next) = buffer.split_at_mut(leftover_len.min(buffer.len()));
-
-            // copy rendered audio into output slice
-            for i in 0..self.channels {
-                let output = first.iter_mut().skip(i).step_by(self.channels);
-                let channel = prev_rendered.channel_data(i)[offset..].iter();
-                for (sample, input) in output.zip(channel) {
-                    let value = Sample::from::<f32>(input);
-                    *sample = value;
-                }
-            }
-
-            // exit early if we are done filling the buffer with the previously rendered data
-            if next.is_empty() {
-                self.buffer_offset = Some((offset + first.len() / self.channels, prev_rendered));
-                return;
-            }
-
-            // if there's still space left in the buffer, continue rendering
-            buffer = next;
-        }
-
-        // The audio graph is rendered in chunks of RENDER_QUANTUM_SIZE frames.  But some audio backends
-        // may not be able to emit chunks of this size.
-        let chunk_size = RENDER_QUANTUM_SIZE * self.channels as usize;
-
-        for data in buffer.chunks_mut(chunk_size) {
-            // handle addition/removal of nodes/edges
-            self.handle_control_messages();
-
-            // update time
-            // @note - this follows the spec as fetch_add returns the old value
-            let timestamp =
-                self.frames_played
-                    .fetch_add(RENDER_QUANTUM_SIZE as u64, Ordering::SeqCst) as f64
-                    / self.sample_rate.0 as f64;
-
-            // render audio graph
-            let rendered = self.graph.render(timestamp, self.sample_rate);
-
-            // copy rendered audio into output slice
-            for i in 0..self.channels {
-                let output = data.iter_mut().skip(i).step_by(self.channels);
-                let channel = rendered.channel_data(i).iter();
-                for (sample, input) in output.zip(channel) {
-                    let value = Sample::from::<f32>(input);
-                    *sample = value;
-                }
-            }
-
-            if data.len() != chunk_size {
-                // this is the last chunk, and it contained less than RENDER_QUANTUM_SIZE samples
-                let channel_offset = data.len() / self.channels;
-                debug_assert!(channel_offset < RENDER_QUANTUM_SIZE);
-                self.buffer_offset = Some((channel_offset, rendered.clone()));
-            }
-        }
-    }
+/// Connection between two audio nodes
+struct Edge {
+    /// index of the current Nodes input/output port
```

Comment:
it is not "output port" only?

Reply:
It can be either an incoming or outgoing edge, so I cannot tell at this point!
But do not despair, I'm cooking up a plan to only store the outgoing edges so this confusion will no longer occur. This does mean I need to change the graph topological sorting algorithm which will take some more time